monai/apps/nuclick/transforms.py

```diff
@@ -0,0 +1,246 @@
+# Copyright (c) MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import math
+import random
+
+import cv2
+import numpy as np
+import skimage
+from skimage.morphology import remove_small_objects
+
+from monai.config import KeysCollection
+from monai.transforms import MapTransform, RandomizableTransform, Transform
+
+
+class FlattenLabeld(MapTransform):
```

Comment:
This seems similar to [`KeepLargestConnectedComponents`](https://github.com/Project-MONAI/MONAI/blob/dev/monai/transforms/post/array.py#L257), except `KeepLargestConnectedComponents` only keeps one instance per labelfield. Could we merge these two together and simply improve `KeepLargestConnectedComponents`? Could have the argument `max_num_instances`, which defaults to `1` for backwards compatibility, but could be set to `-1` to extract all (12 in this example).

Might help us avoid adding `cv2`, too cc @Nic-Ma.

Reply:
Hi @rijobro ,

Sounds like a good idea.
And maybe set to `None` means to extract all?

Thanks.

Reply:
Is `FlattenLabeld` only compatible with binary input? If not, how does multiple labels with each multiple unconnected regions work?

Reply:
@rijobro @Nic-Ma that suggestion makes sense, maybe we can have a new PR to merge this into our existing KeepLargestConnectedComponent (KLCC)?

Once we have it integrated in KLCC, we can remove FlattenLabeld from nuclick transforms, after testing it out with the training pipeline to ensure the results are the same..

Reply:
For what it's worth, I don't really agree with adding a class and then removing it in a follow-up PR as this will break backwards compatibility. I would rather the KLCC classes were merged and generalised straight away.

Reply:
I've tested that using this change (with `skimage.measure`) is non-breaking and it avoids adding new dependencies:
```diff
--- a/monai/apps/nuclick/transforms.py
+++ b/monai/apps/nuclick/transforms.py
@@ -34,7 +34,7 @@ class FlattenLabeld(MapTransform):
     def __call__(self, data):
         d = dict(data)
         for key in self.keys:
-            _, labels, _, _ = cv2.connectedComponentsWithStats(d[key], 4, cv2.CV_32S)
+            labels = measure.label(d[key], connectivity=1).astype(np.int32)
             d[key] = labels.astype(np.uint8)
         return d
```
If you agree, please change the PR, and then we take another PR to refactor this existing logic to be more generic https://github.com/Project-MONAI/MONAI/blob/9e8fbafe0dd73404eeb4b8ad1e06cbe00815adb5/monai/transforms/utils.py#L964-L970 to

Reply:
should be fine to use skimage instead of cv2... not a problem.. u can avoid opencv dep if that's a problem for monai.. but for 2D lots of people use opencv.. for example finding contours is still faster in opencv vs skimage.. however we can revisit this dep in future.. when its really needed in monai..

Reply:
For the FlattenLabeld.. please make sure you verify similar behavior when you switch to KLCC..  across simple label vs multiple label vs un-connected multiple label (e.g. multiple nuclei cells.. of different type.. and each region/patch having more than one count/unconnected ones..) and in that case.. each unconnected one becomes a different label.. for example.. if there are [1, 2, 3] with count (unconnected) of each label type is [2, 1, 4] then after flattening you should get something like [11, 12, 21, 31, 32, 33, 34] with count [1, 1, 1, 1, 1, 1, 1]   (here i am using 11, 12 for easy reference.. to show from where they are coming from..

I guess when vish adds the transform output into to notebook/tutorial, it gets best illustrated.. 

Before:
![image](https://user-images.githubusercontent.com/7339051/170246399-574fc5d2-66a8-439a-98b2-657e3fade0ee.png)

After Flatten:
![image](https://user-images.githubusercontent.com/7339051/170246539-9d90af8b-02d4-48bb-9998-4332d0241630.png)

Reply:
makes sense, I just verified this test case seems to work fine:
```
LABEL_1_1 = np.array(
    [
        [1, 1, 1, 0, 0, 1, 0],
        [1, 1, 1, 0, 0, 1, 0],
        [1, 1, 1, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0],
        [1, 1, 1, 0, 2, 2, 2],
        [1, 1, 1, 0, 2, 2, 2],
        [1, 1, 1, 0, 2, 2, 2],
    ],
    dtype=np.uint8,
)
```
after:
```
array([[1, 1, 1, 0, 0, 2, 0],
       [1, 1, 1, 0, 0, 2, 0],
       [1, 1, 1, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0],
       [3, 3, 3, 0, 4, 4, 4],
       [3, 3, 3, 0, 4, 4, 4],
       [3, 3, 3, 0, 4, 4, 4]], dtype=uint8)
```

Reply:
![sk_image_result](https://user-images.githubusercontent.com/8859225/170350746-d6000e4b-9f20-4d75-9f3b-22eb9fb05c59.png)
![cv2_image_result](https://user-images.githubusercontent.com/8859225/170350849-6329d049-6321-48e0-beba-29ebe87a8622.png)

The skimage result is almost similar to that of cv2. Its 0.791 vs 0.8021 in terms of validation Dice score which can be fairly attributed to random initialization of the network. It should be noted though that there is difference in terms of compute time, skimage implementation is slower as the overall training time was 3 hours vs cv2 implementation where the total time was approximately 2hrs 50 mins.

For now I think it is safe to proceed with the skimage implementation. Thanks a ton @wyli for providing this alternative.